src/ray/common/status_or.h

```diff
@@ -0,0 +1,261 @@
+  // function returns `ray::Ok()`.
+  ABSL_MUST_USE_RESULT const Status &status() const & { return status_; }
+  ABSL_MUST_USE_RESULT Status status() && {
+    Status new_status = std::move(status_);
+    return new_status;
+  }
+
+  // Apply the functor [f] if `this->ok() == true`, otherwise return the error status
+  // contained.
+  template <typename F>
+  auto and_then(F &&f) &;
+  template <typename F>
+  auto and_then(F &&f) const &;
+  template <typename F>
+  auto and_then(F &&f) &&;
+
+  // Returns a reference to the current value.
+  //
+  // REQUIRES: `this->ok() == true`, otherwise the behavior is undefined.
+  //
+  // Use `this->ok()` to verify that there is a current value within the
+  // `ray::StatusOr<T>`. Alternatively, see the `value()` member function for a
+  // similar API that guarantees crashing or throwing an exception if there is
+  // no current value.
+  T &operator*() & { return get(); }
+  const T &operator*() const & { return get(); }
+  T &&operator*() && { return std::move(get()); }
+
+  // Returns a pointer to the current value.
+  //
+  // REQUIRES: `this->ok() == true`, otherwise the behavior is undefined.
+  //
+  // Use `this->ok()` to verify that there is a current value.
+  T *operator->() & { return &data_; }
+  T *operator->() const & { return &data_; }
+
+  // Returns a reference to the held value if `this->ok()`. Otherwise, throws
+  // `std::runtime_error`.
+  //
+  // If you have already checked the status using `this->ok()`, you probably
+  // want to use `operator*()` or `operator->()` to access the value instead of
+  // `value`.
+  //
+  // Note: for value types that are cheap to copy, prefer simple code:
+  //
+  //   T value = statusor.value();
+  //
+  // Otherwise, if the value type is expensive to copy, but can be left
+  // in the StatusOr, simply assign to a reference:
+  //
+  //   T& value = statusor.value();  // or `const T&`
+  //
+  // Otherwise, if the value type supports an efficient move, it can be
+  // used as follows:
+  //
+  //   T value = std::move(statusor).value();
+  //
+  // The `std::move` on statusor instead of on the whole expression enables
+  // warnings about possible uses of the statusor object after the move.
+  T &value() &;
+  const T &value() const &;
+  T &&value() &&;
+
+  void swap(StatusOr &rhs);
+
+  // Copy current value out if OK status, otherwise construct default value.
+  T value_or_default() const & {
+    static_assert(std::is_copy_constructible_v<T>, "T must by copy constructable");
+    static_assert(std::is_default_constructible_v<T>, "T must by default constructable");
+    if (ok()) return get();
+    return T{};
+  }
+  T value_or_default() && {
+    static_assert(std::is_copy_constructible_v<T>, "T must by copy constructable");
+    static_assert(std::is_default_constructible_v<T>, "T must by default constructable");
+    if (ok()) return std::move(get());
+    return T{};
+  }
+
+  static_assert(std::is_default_constructible_v<T>,
```

Comment:
[nit] if we have this, we don't need the same assertion in value_or_default() ?

Reply:
Q: if we use union, maybe we can lift this requirement? but still we need it to be move constructible of course.

Reply:
> if we have this, we don't need the same assertion in value_or_default() ?

Sounds good, updated.

> Q: if we use union, maybe we can lift this requirement?

Maybe? I took a look at abseil status or implementation, it doesn't use union / variant for data and status, let's keep it as it is for now, and let optimization in the followup iteration.

https://github.com/abseil/abseil-cpp/blob/fcc8630eede5498b48b45b99e4eaa1406d6657e9/absl/status/internal/statusor_internal.h#L317-L333